src/base/misc/dos2linux.c

```diff
@@ -528,12 +532,23 @@ static inline int mem_likely_protected(dosaddr_t addr, int len)
      this technique.
    */
   int hash = (addr >> PAGE_SHIFT) & (PAGE_SIZE-1);
-  return unprotected_page_cache[hash] != ((addr + len - 1) & PAGE_MASK);
+  if (unprotected_page_cache[hash] == ((addr + len - 1) & PAGE_MASK))
+    return &unprotected_page_unixaddr_tlb[hash][addr & (PAGE_SIZE-1)];
+  else
+    return NULL;
 }
 
-static inline void set_unprotected_page(dosaddr_t addr)
+/* marks the page as unprotected in the cache, add the host address to
+   the TLB, and returns the host address */
+static inline void *set_unprotected_dosaddr_to_unixaddr(dosaddr_t addr)
 {
-  unprotected_page_cache[(addr >> PAGE_SHIFT) & (PAGE_SIZE-1)] = addr & PAGE_MASK;
+  int hash = (addr >> PAGE_SHIFT) & (PAGE_SIZE-1);
+  unsigned char *uaddr;
+
+  unprotected_page_cache[hash] = addr & PAGE_MASK;
+  uaddr = dosaddr_to_unixaddr(addr & PAGE_MASK);
+  unprotected_page_unixaddr_tlb[hash] = uaddr;
+  return &uaddr[addr & (PAGE_SIZE-1)];
```

Comment:
That func is definitely not the best engineering.
Maybe you mixed so many things in 1 func for
a speed-up. But I think if you move
dosaddr_to_unixaddr() to the caller, and pass
uaddr to that func rather than return it, then
the speed would be the same or even better.

Reply:
So I suggest to keep having set_unprotected_page()
and just pass an uaddr to it as an extra arg.

Reply:
Done

Reply:
For the speed it doesn't really matter. This is all the slow path, the fast path only calls `unprotected_dosaddr_to_unixaddr()`.